### PR TITLE
Enable resetting of batchnorm running stats and cumulative ("simple") moving average

### PR DESCRIPTION
**Pull request by jma127**
This is a proposed implementation of ``reset_running_stats``, which I discussed with @soumith earlier today. In addition, it enables the use of the simple moving average for running stats in place of the exponential moving average, with a ``momentum=None`` setting.

The motivation behind these changes is to faithfully execute the moment estimation of [the original batchnorm paper](https://arxiv.org/abs/1502.03167) (line 10 in Algorithm 2). To do this, you would do:

    bn_layer = BatchNorm(..., momentum=None)
    
    # construct the rest of the network and train
    
    bn_layer.reset_running_stats()
    for i in range(num_bn_meanvar_batches):
        model.forward(get_new_batch())

    bn_layer.eval()

    # use the model

### CODE REVIEW

**Comment by jma127** [comment]
(for some possibly confusing chicken-scratch motivation of why SMA is needed, see https://www.overleaf.com/read/xydtqhmhmtdz)

**Comment by goldsborough** [comment]
Our CI was having issues, but it now works again and the failure is on your side. Please fix your PR :)

**Comment by jma127** [comment]
OK, wasn't sure on that -- thanks for confirming! Looks like I didn't update some JIT unit test expected value, fixed now.

File: torch/nn/modules/batchnorm.py [hunk 1]
```diff
@@ -25,15 +25,22 @@ def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True,
...
-    def reset_parameters(self):
+    def reset_running_stats(self):
         if self.track_running_stats:
             self.running_mean.zero_()
             self.running_var.fill_(1)
+            self.num_batches_tracked.zero_()
+
+    def reset_parameters(self):
+        self.reset_running_stats()
+
```
Comment (SsnL) [inline torch/nn/modules/batchnorm.py]: nit: extra empty line

File: torch/nn/modules/batchnorm.py [hunk 2]
```diff
@@ -93,7 +110,8 @@ class BatchNorm1d(_BatchNorm):
         eps: a value added to the denominator for numerical stability.
             Default: 1e-5
         momentum: the value used for the running_mean and running_var
-            computation. Default: 0.1
+            computation. Can be set to ``None`` for cumulative moving average
+            ("simple" average). Default: 0.1
```
Comment (SsnL) [inline torch/nn/modules/batchnorm.py]: these are indeed simple average, maybe we can remove the quotes? or we can call it arithmetic average?

**Review by ssnl — CHANGES_REQUESTED** [review]
looks good in general except for minor comments. after fixing and rebasing this should be mergeable.

**Comment by SsnL** [comment]
It would also be great if we can test the arithmetic average `test_nn.py`

**Comment by SsnL** [comment]
It seems that there are gloo changes. Could you try `git submodule update --init` to resolve it?

**Comment by jma127** [comment]
Whoops, yeah. There are also some failing tests -- I will debug and @mention you when ready.

**Comment by jma127** [comment]
@SsnL ready for another look!

File: test/test_nn.py [hunk 3]
```diff
@@ -3790,6 +3803,48 @@ def _test_batchnorm_eval(self, test_type=torch.FloatTensor):
...
         self.assertEqual(grad1, grad2)
 
+    def _test_batchnorm_simple_average(self, test_type=torch.FloatTensor):
+        module = nn.BatchNorm1d(3, momentum=None).type(test_type)
+        zeros = torch.zeros(3).type(test_type)
+        ones = torch.ones(3).type(test_type)
+        self.assertEqual(module.running_mean, zeros)
+        self.assertEqual(module.running_var, ones)
+
+        data1 = Variable(torch.rand(4, 3).type(test_type), requires_grad=True)
```
Comment (SsnL) [inline test/test_nn.py]: nit: `Variable` wrapper is not needed anymore. We are also removing it from other test cases, so could you also not use it?

**Review by ssnl — COMMENTED** [review]
Only one minor nit. Otherwise looks great!

**Review by ssnl — APPROVED** [review]
this can be merged if tests pass :) Thanks @jma127 !

**Comment by goldsborough** [comment]
Is the CI not working for you? Tests start automatically (within ~10 secs) every time you push new commits (i.e. Usually no need to use pytorchbot) -- just FYI

**Comment by jma127** [comment]
Oh OK, good to know :P

**Comment by jma127** [comment]
One test is giving a 404 right now so I'll restart again.

**Comment by jma127** [comment]
Hmm, now I get:

    04:51:16 Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.32/containers/4631e92d8a4fd7a9deca52381a4024a52713bcd3979b6425abdff43d7fa91feb?force=1: dial unix /var/run/docker.sock: connect: permission denied

Will retest one more time -- if it still persists, @goldsborough could you investigate?

**Comment by goldsborough** [comment]
Sure

**Comment by goldsborough** [comment]
Let's merge it
